### ISSUE TITLE
Y'all got some strategy for mocking ActiveRecord associations? 

### ISSUE THREAD

**Issue by dhnaranjo** [testdouble/mocktail#21]
I have a method on an AR model that calls a method on an associated model and does some biz with the result. I'd like to mock it but I can't seem use a Mocktail to replace an ActiveRecord association. Y'all got thoughts?

So here's my lil example script:

```
# frozen_string_literal: true
require 'bundler/inline'
require 'minitest/autorun'
require 'mocktail'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'rails', '~> 7.0.4'
  gem 'sqlite3'
  gem 'mocktail'
end

require 'rails/all'
database = 'development.sqlite3'

ENV['DATABASE_URL'] = "sqlite3:#{database}"
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: database)
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class MocktailTest < ActiveSupport::TestCase
  include Mocktail::DSL

  test "associating a mocktail" do
    post = Mocktail.of(Post)
    Comment.new(post: post)
  end
end
```

which when ran hits me with a:
```
Minitest::UnexpectedError: NoMethodError: undefined method `primary_key' for #<Class for mocktail of Post:0x0000000106e3b4b0>
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/reflection.rb:317:in `primary_key'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/reflection.rb:744:in `association_primary_key'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:130:in `primary_key'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:122:in `replace_keys'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:93:in `replace'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/singular_association.rb:19:in `writer'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/builder/association.rb:112:in `post='
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:49:in `public_send'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:49:in `_assign_attribute'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:21:in `block in _assign_attributes'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:13:in `each'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:13:in `_assign_attributes'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:34:in `assign_attributes'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/core.rb:468:in `initialize'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/inheritance.rb:75:in `new'
    /Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/inheritance.rb:75:in `new'
    /Users/dazmin/RubymineProjects/mocktail_rails_test/mocktail_rails_test.rb:42:in `block in <class:MocktailTest>'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/reflection.rb:317:in `primary_key'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/reflection.rb:744:in `association_primary_key'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:130:in `primary_key'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:122:in `replace_keys'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/belongs_to_association.rb:93:in `replace'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/singular_association.rb:19:in `writer'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/associations/builder/association.rb:112:in `post='
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:49:in `public_send'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:49:in `_assign_attribute'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:21:in `block in _assign_attributes'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:13:in `each'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/attribute_assignment.rb:13:in `_assign_attributes'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activemodel-7.0.4.3/lib/active_model/attribute_assignment.rb:34:in `assign_attributes'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/core.rb:468:in `initialize'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/inheritance.rb:75:in `new'
/Users/dazmin/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/activerecord-7.0.4.3/lib/active_record/inheritance.rb:75:in `new'
/Users/dazmin/RubymineProjects/mocktail_rails_test/mocktail_rails_test.rb:42:in `block in <class:MocktailTest>'

Process finished with exit code 1
```

**Comment by searls** [comment]
**tl;dr I recommend not doing this, and Mocktail is never going to bend over backwards to support things like this, because tests that do so are typically counter-productive.**

Hi @dhnaranjo! My advice is to not mock ActiveRecord objects specifically, or [anything you don't own](https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own) generally. And even though you might have defined the subclass of a controller or model, you really only own the tip of its behavioral iceberg. 99% of the complexity lives below the surface, and as a result its design is more or less fixed in stone. Since mocking libraries only exist to provide design feedback to the tester (by way of the fact that easy-to-orchestrate-via-mocking things are typically easy-to-use things), any feedback you get that Rails types are hard to mock is useless: not like you can fix or improve Rails.

Additionally, I would always recommend separating the units that contain application's feature behavior from the units that represent data / values. The data flows through the units that implement that feature logic as arguments and return values. Values themselves tend to be more or less like structs, free of much meaningful logic themselves, apart from elucidative methods (say `full_name` that combines `first_name` and `last_name` fields) that make them more pleasant when consumed.

Finally, and when you combine the above two design principles, I only ever  would only ever recommend **mocking the dependencies of units that implement application behavior**, which themselves are other units that implement application behavior. If you implement all your application behavior inside objects that extend from Rails superclasses, then I don't recommend mocking at all, since the feedback such tests can provide will only be useless, futile pain -- if you encounter a painful API interaction, it's not like rails-core is going to accept a patch to change it.

This may all seem really dense if you're not familiar with discussions of the above topics (and you're not alone, >99% of developers are not), you might enjoy ]this conference talk](https://blog.testdouble.com/talks/2018-03-06-please-dont-mock-me/) where I discuss a lot of these issues

**Comment by dhnaranjo** [comment]
I get that argument, and I got here because of your talks which I enjoyed, but since you announced this thing at a RailsConf I think it'd be worth noting this philosophically driven incompatibility with ActiveRecord. I spent a good amount of time tryinna figure out what I was doing wrong.

I don't supposed you'd be amenable to a PR to the readme stating the incompatibility so future folks might save some frustration?

**Comment by searls** [comment]
Yeah, I appreciate your frustration and I'm sorry for it. While in my head this isn't incompatible with ActiveRecord at all (because AR models are often the things being passed to and returned from my mocks), I get where you're coming from.

I had the same thought, that the README should indicate where and when it's designed to work. This should probably be part of a broader change to the readme to reposition the tool as intentionally opinionated.

In the short term I'd definitely accept a PR to that effect, that Mocktail is designed to replace code that you own, as opposed to objects you extend from other gems like activerecord.
